### PR TITLE
Draft of how to fix compiler flags

### PR DESCRIPTION
**Pull request by jjwilke**
Not in any way done...

We have the problem of many, many flags that are sometimes direct flags to a host compiler and sometimes `-Xcompiler` flags passed through a device compiler to a host compiler. This is the best I could come up with (for now) for how to do this cleanly.

The two main questions are:
1. General interface
2. Some builds are now going to require CMake 3.13 if you activate architecture flags (this is optional, you can always deactivate the architecture flags). This is because of annoying flag deduplication issues

### CODE REVIEW

**Comment by masterleinad** [comment]
Is there a specific problem this pull request tries to solve?

File: cmake/kokkos_functions.cmake [hunk 1]
```diff
@@ -825,6 +827,43 @@ FUNCTION(COMPILER_SPECIFIC_OPTIONS_HELPER)
   ENDIF()
 ENDFUNCTION(COMPILER_SPECIFIC_OPTIONS_HELPER)
 
+FUNCTION(COMPILER_SPECIFIC_OPTIONS_HELPER_HELPER)
```
Comment (masterleinad) [inline cmake/kokkos_functions.cmake:830]:
```suggestion
FUNCTION(COMPILER_SPECIFIC_OPTIONS_HELPER)
```

File: cmake/kokkos_functions.cmake [hunk 2]
```diff
@@ -769,47 +769,49 @@ FUNCTION(kokkos_link_tpl TARGET)
...
-    "COMPILER_ID"
-    "${COMPILERS}"
+    "PREFIX;COMPILER_ID"
+    "${COMPILER}"
     ${ARGN})
+
+  FOREACH(COMP ${COMPILERS})
+    IF (NOT PARSE_${COMPILER})
+      SET(PARSE_${COMPILER} ${PARSE_DEFAULT})
+    ELSE IF("${PARSE_${COMPILER}" STREQUAL "NO-VALUE-SPECIFIED")
```
Comment (masterleinad) [inline cmake/kokkos_functions.cmake:786]:
```suggestion
    ELSEIF("${PARSE_${COMPILER}" STREQUAL "NO-VALUE-SPECIFIED")
```

File: cmake/kokkos_functions.cmake [hunk 3]
```diff
@@ -769,47 +769,49 @@ FUNCTION(kokkos_link_tpl TARGET)
...
+FUNCTION(COMPILER_SPECIFIC_OPTIONS_HELPER_HELPER)
+  SET(COMPILERS PGI XL DEFAULT Cray Intel Clang AppleClang IntelClang GNU NVIDIA HIP)
+
   CMAKE_PARSE_ARGUMENTS(
     PARSE
     "LINK_OPTIONS;COMPILE_OPTIONS;COMPILE_DEFINITIONS;LINK_LIBRARIES"
-    "COMPILER_ID"
-    "${COMPILERS}"
+    "PREFIX;COMPILER_ID"
+    "${COMPILER}"
```
Comment (masterleinad) [inline cmake/kokkos_functions.cmake:780]:
```suggestion
    "${COMPILERS}"
```

**Review by masterleinad — COMMENTED** [review]
With the proposed changes, this looks good to me but I don't have a good idea how to handle the deduplicated flags yet either.

**Comment by jjwilke** [comment]
Sorry, should have added more detail.  We have places like:

````cmake
IF (KOKKOS_ARCH_POWER8)
  COMPILER_SPECIFIC_FLAGS(
    PGI     NO-VALUE-SPECIFIED
    NVIDIA  NO-VALUE-SPECIFIED
    DEFAULT -mcpu=power8 -mtune=power8
  )
ENDIF()
````
Even if we are using nvcc, though, we actually do want the architecture flags passed through to the host compiler using `-Xcompiler`.  We need a way of distinguishing:

1. This flag is only valid for host compilers. Don't include with a device compiler
2. This flag is valid for host and device compilers, but needs to be prefixed with`-Xcompiler` and passed through
3. This flag is valid for a device compiler with no extra prefix

**Pull request closed without merging**